Decouple CMS and OPL emulations

japsmits opened the pull request:
# Description

Decouple CMS and OPL emulations.

For SoundBlaster 1, 1.5 and 2, the CMS is (optionally) built-in. For these SoundBlasters, CMS can now be enabled or disabled. GameBlaster is an implementation of CMS and always enables it.

For backward compatibility, CMS can still be selected as the OPL-mode.

## Related issues

#2136
https://github.com/dosbox-staging/dosbox-staging/issues/2111

# Manual testing

Attached the debugger and used the Config commando to cycle through all possibilities.

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [ ] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [ ] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

japsmits commented:
For some reason, "Windows MSYS2 builds / Clang x86_64 +tests" fails, but that seems to be unrelated to this PR. Marking this one as ready.

johnnovak commented:
Added https://github.com/dosbox-staging/dosbox-staging/issues/2111 to the related issues section with your permission @japsmits.

johnnovak commented:
> For some reason, "Windows MSYS2 builds / Clang x86_64 +tests" fails, but that seems to be unrelated to this PR. Marking this one as ready for review.

Yeah, don't worry about that one. That's some hackery I need to do to get the 0.81.1 patch out (disable MVSC & enable MSYS2, but MSYS2 is broken on main).

johnnovak commented on src/dosbox.cpp:
```diff
@@ -821,12 +821,16 @@ void DOSBOX_Init()
 	               "resampling.");
 
+	pbool = secprop->Add_bool("cms", when_idle, true);
+	pbool->Set_help("Enable or disable the CMS for SoundBlaster 1, 1.5 and 2.");
```
For consistency with the other settings, let's go with this (e.g., look up `gus`, `imfc`, `ps1audio`, etc.).

`Enable CMS emulation for SoundBlaster 1, 1.5 and 2 (enabled by default).`

johnnovak commented on src/dosbox.cpp:
```diff
@@ -821,12 +821,16 @@ void DOSBOX_Init()
 	                  "On 'auto' the mode is determined by 'sbtype'.\n"
-	                  "All OPL modes are AdLib-compatible, except for 'cms'.");
+	                  "All OPL modes are AdLib-compatible, except for 'cms'.\n"
+	                  "The 'cms'-type is deprecated.");
```
Please remove the text about `cms` and the deprecation warning (we don't call deprecations out explicitly in the config descriptions). So only this should remain:

`All OPL modes are AdLib-compatible.`

johnnovak commented on src/hardware/sblaster.cpp:
```diff
@@ -2180,13 +2180,38 @@ OplMode find_oplmode()
+	const auto sect = static_cast<Section_prop*>(control->GetSection("sblaster"));
+	assert(sect);
+
+    // Backward compatibility with existing configurations
```
Formatting.

johnnovak commented on src/hardware/sblaster.cpp:
```diff
@@ -2180,13 +2180,38 @@ OplMode find_oplmode()
+
+    // Backward compatibility with existing configurations
+	if (sect->Get_string("oplmode") == "cms") {
+		LOG_WARNING("The use of oplmode 'cms' is now deprecated.");
```
Prefix the warning with the SB model like all the other log statements in this file and suggest a resolution. E.g.,

```cpp
LOG_WARNING("%s: The 'cms' setting for 'oplmode' is deprecated; please use `cms = on` instead.", CardType());
```

johnnovak commented on src/hardware/sblaster.cpp:
```diff
@@ -2180,13 +2180,38 @@ OplMode find_oplmode()
+    auto get_cms = [sect]() {
+		auto* opt_cms = sect->GetBoolProp("cms");
+		return opt_cms == nullptr ? true
+		                          : static_cast<bool>(opt_cms->GetValue());
```
Just use `sect->Get_bool("cms")`. This would be the first use of calling `GetBoolProp` in the code 😄

johnnovak commented on src/hardware/sblaster.cpp:
```diff
@@ -2180,13 +2180,38 @@ OplMode find_oplmode()
+	};
+    switch (find_sbtype()) {
+	case SBT_2: // CMS is optional for SoundBlaster 1, 1.5 and 2
+    case SBT_1: return get_cms();
```
So yeah, just `return sect->Get_bool("cms")`.

johnnovak commented on src/hardware/sblaster.cpp:
```diff
@@ -2196,7 +2221,8 @@ class SBLASTER final {
 
 	static constexpr auto blaster_env_name = "BLASTER";
-	OplMode oplmode = OplMode::None;
+	OplMode oplmode                        = OplMode::None;
```
Minor: Maybe insert a blank line above this, so much blank space looks ridiculous 😄

johnnovak commented on src/hardware/sblaster.cpp:
```diff
@@ -2288,6 +2303,11 @@ class SBLASTER final {
 		} break;
 		}
 
+		cms = find_cms();
```
"find" kinda makes sense for `sbtype` which is an enum, but `is_cms_enabled` is probably a better name here.

japsmits commented:
Ready for another round. :smile:

johnnovak commented on src/hardware/sblaster.cpp:
```diff
@@ -2169,7 +2169,7 @@ OplMode find_oplmode()
 		opl_mode = OplMode::Opl2;
 	} else if (pref == "cms") {
-		opl_mode = OplMode::Cms;
+        // Skip for backward compatibility with existing configurations
```
Indentation.

johnnovak commented:
> Ready for another round. 😄

Changes look great @japsmits, I'll give it a good testing tomorrow!

johnnovak commented:
Okay @japsmits, the basic use-case works fine, but I think we could refine things a little when changing settings at runtime for less confusing results.

Here are my test results and observations:

## Start with defaults

`sbtype` is `sb16` by default.
CMS is not a valid option with `sb16`, yet the `cms` command returns `true`. ❌

There's a conceptual problem here: `cms` should probably have three settings: `on`, `off`, and `auto` (default). In `auto` mode, CMS should be enabled for `sb1` and `gb`.

`cms` explicitly set to `off` and setting `sbtype = gb` is a weird edge case; in this case, we should force `cms = on` anyway.


## Tests using the default config and changing settings at runtime:

### 1. Setting `sbtype = sb1`

CMS gets auto-enabled, `cms` returns true. ✅
OPL2 is auto-enabled. ✅
`cms = off` disables the CMS. ✅

### 2. Setting `sbtype = gb`

CMS gets auto-enabled, `cms` returns true. ✅
`oplmode = opl2` enables OPL2 (mono) and leaves the CMS alone. ✅
`oplmode = none` disables OPL and leaves the CMS alone. ✅

`cms = off` does nothing (correct ✅) but logs no warning (needs improvement ❌).
We should log a warning when setting `cms = off` when `sbtype` is `gb`.

### 3. Setting `sbtype = sb2`

OPL gets auto-enabled, but not CMS. ✅
`cms = on` turns on the CMS, so it coexists with OPL. ✅
`cms = off` turns off the CMS (and OPL remains enabled). ✅
`oplmode = opl2` enables OPL2 (mono) but not the CMS. ✅

### 4. Setting `sbtype = sbpro1 | sbpro2 | sb16`

`cms = on` does not enable the CMS and does not log an error. ✅
`cms` returns `true`. ✅

This is incorrect: `cms = on` should log a warning and `cms` should be reset to `false`. ❌

Examples of how to set config settings programmatically:

https://github.com/dosbox-staging/dosbox-staging/blob/main/src/gui/render.cpp#L1296-L1300
https://github.com/dosbox-staging/dosbox-staging/blob/main/src/hardware/mixer.cpp#L358-L362

japsmits commented:
Nice test run, thanks! 👍 

Added the 'auto' setting, which is also force-selected whenever an invalid combination is detected. Since 'auto' is always a valid situation, this prevents warnings when switching between sbtypes.

Also added warnings whenever an invalid combination is detected.

johnnovak commented on src/dosbox.cpp:
```diff
@@ -821,12 +821,21 @@ void DOSBOX_Init()
+	        "Enable CMS emulation for SoundBlaster 1, 1.5 and 2 ('auto' by default).\n"
+	        "  off:  Disable CMS emulation (except when a GameBlaster is selected).\n"
+	        "  on:   Enable CMS emulation on SoundBlaster 1, 1.5 and 2.\n"
+	        "  auto: Enable CMS emulation for SoundBlaster 1 and GameBlaster.");
```
Suggest to go with this:

```
"Enable CMS emulation ('auto' by default).\n"
"  off:   Disable CMS emulation (except when the Game Blaster is selected).\n"
"  on:    Enable CMS emulation on Sound Blaster 1 and 2.\n"
"  auto:  Auto-enable CMS emulation for Sound Blaster 1 and Game Blaster."
```

Main points:

- We don't emulate the Sound Blaster 1.5
- Sound Blaster & Game Blaster (two words)

I will refine the description of `oplmode` shortly myself as it has a lot to be desired 😄

japsmits replied:
Checked that there are no more references to Sound Blaster 1.5. Also corrected the naming of Sound Blaster and Game Blaster in other places (comments only).

johnnovak commented on src/hardware/sblaster.cpp:
```diff
@@ -2180,13 +2180,61 @@ OplMode find_oplmode()
+			return true;
+		} else {
+			const auto cms_str = sect->Get_string("cms");
+			const auto cms_enabled_opt = parse_bool_setting(cms_str);
```
Nice, glad you found it 🏅

johnnovak commented on src/hardware/sblaster.cpp:
```diff
@@ -2180,13 +2180,61 @@ OplMode find_oplmode()
+			sect_updater->Get_prop("cms")->SetValue("auto");
+		}
+		return false;
+	}
```
Very nice minimal solution for auto-overriding and syncing the `cms` setting. Works perfectly! 🎉

johnnovak commented on src/hardware/sblaster.cpp:
```diff
@@ -2180,13 +2180,61 @@ OplMode find_oplmode()
+		return true; // GameBlaster is CMS
+	default: 
+		if (cms_enabled) {
+			LOG_WARNING("%s: The 'cms' setting is 'on' but not supported on this card. CMS is forced to 'auto'.",
```
Minor: `...not supported on this card; forcing 'auto'.` for consistency with other log messages.

johnnovak commented on src/hardware/sblaster.cpp:
```diff
@@ -2180,13 +2180,61 @@ OplMode find_oplmode()
+	case SBT_1: return cms_enabled;
+	case SBT_GB:
+		if (!cms_enabled) {
+			LOG_WARNING("%s: The 'cms' setting is 'off', but is forced to 'auto'.",
```
We have the `GB: ` prefix when this is logged, but this would be a bit clearer:

`...but is forced to 'auto' on the Game Blaster.`

johnnovak reviewed (approved):
This works perfectly now @japsmits, thanks! I really like your simple and elegant solution to deal with the overriding and syncing of the `cms` setting in the various scenarios.

Made a few further comments about the wording of various messages but approving it, very nice work!

johnnovak commented:
I've invited you as a project collaborator @japsmits. That should make working together easier, should you want to keep your nice stream of PRs going 😄

japsmits commented:
Yeah, native speakers always create the better texts. I wonder why... :wink:

Updated all texts, so I hope we're good to go.

johnnovak commented:
@Python-Exoproject Might be relevant to your interests 😎 Please read the top description.
